datasets.py

```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
+    padded_shapes = {}
+    for feature_name in feature_names:
+        num_channels = _get_spoken_digit_feature_channels(feature_name)
+        signature[feature_name] = tf.TensorSpec(shape=(None, num_bins, num_channels), dtype=tf.float32)
+        padded_shapes[feature_name] = (None, num_bins, num_channels)
+        
+    signature = [signature]
+        
+    if load_labels:
+        signature.append(tf.TensorSpec(shape=(), dtype=tf.int8))
+        padded_shapes = (padded_shapes, ())
+    
+    return tuple(signature), padded_shapes
+
+
+def _extract_spoken_digit_features_npy(example, config, load_labels):
+    audio = example.audio # this should be a numpy array
+    label = example.label if load_labels else None
+    # TODO data augmentation
+
+    windowed_frame_config = config['dsp']['windowed_frames']
+    stft_features_config = config['dsp']['stft_features']
+    
+    windowed_frame_data = dsp.windowed_frame_analysis(audio, windowed_frame_config)
+    features = dsp.stft_analysis(windowed_frame_data['frames'], stft_features_config)
+    
+    # convert numpy arrays to tensors
+    for feature_name in features:
+        num_channels = _get_spoken_digit_feature_channels(feature_name)
+        feature_frames = features[feature_name].view(f'({num_channels},)float')
+        features[feature_name] = tf.convert_to_tensor(feature_frames, dtype=tf.float32)
+    
+    if label is not None:
+        label = tf.convert_to_tensor(label, dtype=tf.int8)
+        return (features, label)
+    
+    return features
+
+
+def _count_spoken_digit_dataset_npy(data_dir):
+    return sum([1 for _ in data_dir.glob('*.npy')])
+
+
+def _load_spoken_digit_dataset_npy(data_dir, config, load_labels):
+    for data_file in data_dir.glob('*.npy'):
+        data = np.load(data_file)
+        yield _extract_spoken_digit_features_npy(np.rec.array(data, dtype=data.dtype), config, load_labels)
+
+
+def load_spoken_digit_dataset(data_dir, config, load_labels):
+    count = _count_spoken_digit_dataset_npy(data_dir)
+    generator = lambda: _load_spoken_digit_dataset_npy(data_dir, config, load_labels)
+    out_signature, padded_shapes = _get_spoken_digit_signature_and_padded_shape(config, load_labels)
+    all_data = tf.data.Dataset.from_generator(generator, output_signature=out_signature)
+    all_data = all_data.apply(tf.data.experimental.assert_cardinality(count))
+    return all_data, padded_shapes
+
+
+def create_spoken_digit_datasets(
+        data_dir,
+        config,
+        batch_size=32,
+        train_test_split=0.8,
+        load_labels=False,
+        repeat=False,
+) -> Tuple[tf.data.Dataset, tf.data.Dataset]:    
+    all_data, padded_shapes = load_spoken_digit_dataset(data_dir, config, load_labels)
+    total_count = len(all_data)
+
+    train_count = int(train_test_split * total_count)
+    test_count = total_count - train_count
+    
+    train_data = all_data.take(train_count).cache().shuffle(train_count, reshuffle_each_iteration=True)\
```

Comment:
The `.take/ .skip` way works, but there are flaws with it. Notably, when the data is sorted in a particular way, say by speaker or by digit in our case, we'll end up with a test set which may consist only a subset of digits or speakers, while that same digit or speaker would be completely missing from the train set. Note that this is also because we don't want to shuffle before splitting, especially when the shuffle isn't reproducible. 

I think a good way here would be to allow `load_spoken_digit_dataset` and `_load_spoken_digit_dataset_npy` to generate test and train datasets. For example, it could iterate over files in `data_dir` and take every other `n` samples for the test set, and the rest for train. This is what I had done with the `skip_condition` in the code you have removed.

Reply:
Also one motivation for saving things as numpy files was to provide more flexibility.

Reply:
I assume there's some mechanism in TF that would be better than putting the `if n % N` kind of thing in the generator. What if we just use a fixed random seed and shuffle the full dataset before the `.take` / `.skip` ?

Reply:
i've added a deterministic shuffle to the raw TF dataset prior to writing the examples as numpy files